### PR TITLE
8260517: implement Sealed Classes as a standard feature in Java

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
Please review this PR that intents to make sealed classes a final feature in Java. This PR contains compiler and VM changes. In line with similar PRs, which has made preview features final, this one is mostly removing preview related comments from APIs plus options in test cases. Please also review the related [CSR](https://bugs.openjdk.java.net/browse/JDK-8265090)

Thanks

note: this PR is related to [PR-3528](https://github.com/openjdk/jdk/pull/3528) and must be integrated after it.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260517](https://bugs.openjdk.java.net/browse/JDK-8260517): implement Sealed Classes as a standard feature in Java

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 6e2a99c62dca51aa5bbc3df2bd57cb33b8a91283
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to 8ebe56fd65f5aeddc97c4397f28590659e9f46b4
 * [Jan Lahoda](https://openjdk.java.net/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to 304fd76af69a50828a2e6d7baccdab96b08ce4bb

### Contributors
 * Harold Seigel `<hseigel@openjdk.org>`
 * Vicente Romero `<vromero@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3526/head:pull/3526` \
`$ git checkout pull/3526`

Update a local copy of the PR: \
`$ git checkout pull/3526` \
`$ git pull https://git.openjdk.java.net/jdk pull/3526/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3526`

View PR using the GUI difftool: \
`$ git pr show -t 3526`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3526.diff">https://git.openjdk.java.net/jdk/pull/3526.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/classFileParser.cpp [hunk 1]
```diff
@@ -3919,8 +3913,8 @@ void ClassFileParser::parse_classfile_attributes(const ClassFileStream* const cf
             parsed_record_attribute = true;
             record_attribute_start = cfs->current();
             record_attribute_length = attribute_length;
-          } else if (tag == vmSymbols::tag_permitted_subclasses()) {
-            if (supports_sealed_types()) {
+          } else if (_major_version >= JAVA_17_VERSION) {
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classFileParser.cpp:3916]: Can you update to the comment at L3932 to say JAVA_17_VERSION please.

File: src/hotspot/share/classfile/classFileParser.hpp [hunk 2]
```diff
@@ -341,9 +341,6 @@ class ClassFileParser {
                                       const u1* const record_attribute_start,
                                       TRAPS);
 
-  bool supports_sealed_types();
-  bool supports_records();
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/classFileParser.hpp:345]: Good catch!

**Review by dholmes-ora — APPROVED** [review]
Hi Vicente,

Hotspot and hotspot tests all look fine. One query: why was this test removed?

 test/hotspot/jtreg/runtime/sealedClasses/AbstractSealedTest.java

is that functionality tested elsewhere? (The other deleted test seemed obviously trivial.)

Thanks,
David

**Comment by vicente-romero-oracle** [comment]
> Hi Vicente,
> 
> Hotspot and hotspot tests all look fine. One query: why was this test removed?
> 
> test/hotspot/jtreg/runtime/sealedClasses/AbstractSealedTest.java
> 
> is that functionality tested elsewhere? (The other deleted test seemed obviously trivial.)
> 
> Thanks,
> David

Hi David, thanks for your comments, yes regarding `test test/hotspot/jtreg/runtime/sealedClasses/AbstractSealedTest.java`, it was removed because the functionality is tested in `test/langtools/tools/javac/sealed/SealedCompilationTests.java`

File: src/hotspot/share/classfile/classFileParser.cpp (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/hotspot/share/classfile/classFileParser.cpp:3916]: sure

**Review by mcimadamore — APPROVED** [review]
Compiler changes look good (I have not checked SymbolGenerator).

Why were some tests removed?

**Comment by vicente-romero-oracle** [comment]
> Compiler changes look good (I have not checked SymbolGenerator).
> 
> Why were some tests removed?

thanks for the review. The removed tests were already covered in langtools regression tests, so I only removed duplicated tests

**Review by lahodaj — APPROVED** [review]
javac changes look good to me.

**Pull request closed without merging**